### DISCUSSION TITLE
Combining constraints with intersection type

### DISCUSSION THREAD

**Discussion by mcallisto** [Iltotore/iron#27] [Q&A] [answered]
Is there a plain way to combine constraints?

I tried with intersection type
`type +>[A, B] = A ==> Greater[B] & Even[A]`

but when testing I get

```
Found:    (2 : Int)
Required: Int +> (0 : Int)
```

Even if `import iron.constraint.refineValue` is available.

**Comment by Iltotore** [comment, marked as answer]
Hello,
the use of intersection here is impossible in its current state because intersection is commutative leading to an implicit ambiguity error (multiple possibilities).

Instead, you can use the `&&[A, B]` alias of `And`:
```scala
type +>[A, B] = A ==> Greater[A] && Even[B]
``` 

EDIT: Marked as answer. Check the below comments.

**Reply by mcallisto** [reply]
Thank you for the quick reply.

With the change you suggested, now in my `foo` method

```
import iron.constraint.refineValue

def foo(x: Int +> 0, y: Int > 0): Either[IllegalValueError[?], Bar] =
  for
    rx <- x
    ry <- y
  yield
      ...
```
I get the following error
```
value flatMap is not a member of Int +> (0 : Int), but could be made available as an extension method.

The following import might make progress towards fixing the problem:

  import io.github.iltotore.iron.constraint.refineValue


      rx <- x
```
and the same goes for `map` if I move `rx <-x` as second term of the _for expression_.

**Reply by Iltotore** [reply]
This is because `Int +> 0` isn't an Either directly but can be converted to. You have to import `io.github.iltotore.iron.*` or `io.github.iltotore.constrainedToValue`.

Here are [the recommended imports](https://github.com/Iltotore/iron/wiki/Constraint-Usage)

**Reply by mcallisto** [reply]
Sorry, I omitted it in the snippet but I was already using the recommended imports:

```
import io.github.iltotore.iron.*
import io.github.iltotore.iron.constraint.*
import io.github.iltotore.iron.numeric.constraint.{Even, Greater, given}
import io.github.iltotore.iron.constraint.refineValue

type >[A, B] = A ==> Greater[B]
type +>[A, B] = A ==> Greater[B] && Even[A]

trait Foo {

  def foo(x: Int +> 0, y: Int > 0): Either[IllegalValueError[?], Bar] =
    for
      rx <- x
      ry <- y
    yield
      ...

  ...
}
```

Something missing or should I try a manual conversion?

**Reply by Iltotore** [reply]
I mean you should import `given` and not only `*`.

```scala
import io.github.iltotore.iron.{given, *}
```

See https://docs.scala-lang.org/scala3/reference/contextual/given-imports.html

**Reply by mcallisto** [reply]
My bad, you are right, forgot about the Scala 3 `given` imports rule.

But still cannot get rid of the error.
I made a reproducer on Scastie:
https://scastie.scala-lang.org/mcallisto/k2xv7OZZRISl6GhIeH4yCQ/1

Can you have a look?

**Reply by Iltotore** [reply]
You forgot the parantheses on your +> alias.

https://scastie.scala-lang.org/q9Umc8gQRaWjI9LdW5LgIg

It appears that the Even constraint doesnt work in this context and I replaced it in your example by Divisible[2]. I will investigate this issue.

**Reply by mcallisto** [reply]
Thank you very much, it is working now!

While testing with ScalaTest, I noticed the following strange (to me) behavior:

```
   // testing type >[A, B] second argument of foo
    foo(2, 1).isRight shouldBe true // success
    "foo(2, 0)" shouldNot compile   // success

   // testing type +>[A, B] first argument of foo
    foo(4, 1).isRight shouldBe true // success
    foo(3, 1).isLeft shouldBe true  // success, but I would have expected compile-time failure
    "foo(3, 1)" should compile      // success, but I would have expected compile-time failure
    "foo(3, 0)" shouldNot compile   // success, this fails at compile-time but because of invalid second argument

```

**Reply by Iltotore** [reply]
This is the correct behaviour. If the constraint result is fully inlined, and if rules allow it, the result will be evaluated at compile-time.

You can prevent this behaviour using a RuntimeOnly constraint. Currently, there is no out-of-the-box way to transform but this can be achieved pretty easily: https://scastie.scala-lang.org/UGlq85uKTsu33lCjslvFrQ

- Note 1: `type >[A, B] = A ==> Greater[B]` already exists in the [numeric module](https://iltotore.github.io/iron/scaladoc/api/io/github/iltotore/iron/numeric/constraint$.html)
- Note 2: The `RuntimeOnly` modifier shown in the above sample should be useful in multiple cases. I think I will add it for the next release of Iron. Thank your feedback!

I'm still open if you are other questions.

**Reply by mcallisto** [reply]
Thank you for your kind explanations!

But what if, the other way round, I would like `type +>[A, B]` to be evaluated at compile time? Is it possible? I think rules should allow it. How to achieve this?

**Reply by Iltotore** [reply]
Infortunately, the And constraint cannot be inline due to a limitation of Dotty on inline implicit parameters (altrough I didn't test with 3.0.1).

**Reply by mcallisto** [reply]
I see.

But could it be achieved with a custom constraint as shown in https://github.com/Iltotore/iron/wiki/Constraint-Creation#with-parameters? With something like:

`def assert(value: A): Boolean = NumberOrdering.gt(value, constValue[V]) && modulo(value, 2) == 0`

**Reply by Iltotore** [reply]
It's not the same "parameter" on my last message I mean "implicit parameters". Not generic types used as parameters

**Reply by mcallisto** [reply]
This custom constraint added to my repo works for me, failing at compile time as well

```
package io.github.iltotore.iron.numeric

import io.github.iltotore.iron.==>
import io.github.iltotore.iron.constraint.Constraint

import scala.compiletime.{constValue, summonInline}

object constraintCustom {

  /**
   * Constraint: checks if the input value is even and greater than V.
   * @tparam V
   */
  trait EvenAndGreater[V]

  type +>[A, B] = A ==> EvenAndGreater[B]

  class EvenAndGreaterConstraint[A <: Number, V <: A] extends Constraint[A, EvenAndGreater[V]] {
    override inline def assert(value: A): Boolean = NumberOrdering.gt(value, constValue[V]) && modulo(value, 2) == 0

    override inline def getMessage(value: A): String = "value should be even and greater than the specified value"
  }

  inline given[A <: Number, V <: A]: EvenAndGreaterConstraint[A, V] = new EvenAndGreaterConstraint
  
}
```

**Reply by Iltotore** [reply]
That's actually normal 😋 

Your constraint doesn't require implicit parameters. As you can see in the `And` implementation:
```scala
  class AndConstraint[A, B, C, CB <: Constraint[A, B], CC <: Constraint[A, C]](using left: CB, right: CC) extends Constraint.RuntimeOnly[A, And[B, C]] {

    override inline def assert(value: A): Boolean = left.assert(value) && right.assert(value)

    override inline def getMessage(value: A): String = s"${left.getMessage(value)} and ${right.getMessage(value)}"
  }
```

`left` and `right` can't be `inline` due to a compiler limitation leading to deferrence (at least in 3.0.0).

**Reply by mcallisto** [reply]
I guessed it was normal, but in this way I achieved what I needed :sweat_smile: , that is a combined constraint failing at compile time! Thank you for all your help!

**Reply by Iltotore** [reply]
You're welcome 😄
